lib/stats.dx

```diff
@@ -0,0 +1,243 @@
+
+data GeometricDist = Geometric Float
+
+instance Dist GeometricDist Nat
+  log_pdf = \(Geometric p) n.
+    nf = n_to_f n
+    ((nf-1)*(log1p (-p))) + (log p)
+  log_cdf = \(Geometric p) n.
+    nf = n_to_f n
+    log1p (-pow (1-p) nf)
+  log_sf = \(Geometric p) n.
+    (n_to_f n) * (log1p (-p))
+  quantile = \(Geometric p) q.
+    f_to_n $ ceil $ (log1p (-q)) / (log1p (-p))
+  draw = \(Geometric p) k.
+    f_to_n $ ceil $ (log1p (-(rand k))) / (log1p (-p))
+
+'### Poisson distribution
+The [Poisson distribution](https://en.wikipedia.org/wiki/Poisson_distribution) is parameterised by its rate, `r > 0`. Mean is `r`.
+
+data PoissonDist = Poisson Float
+
+def poisson_log_pdf_private (r: Float) (n: Nat) : Float =
+  nf = n_to_f n
+  (nf * (log r)) - r - lgamma(nf + 1)
+
+instance Dist PoissonDist Nat
+  log_pdf = \(Poisson r) x.
+    poisson_log_pdf_private r x
+  log_cdf = \(Poisson r) x.
+    xp1:Nat = x + 1
+    logsumexp $ for i:(Fin xp1). poisson_log_pdf_private r (ordinal i)
+  log_sf = \(Poisson r) x.
+    xp1:Nat = x + 1
+    lcdf = logsumexp $ for i:(Fin xp1). poisson_log_pdf_private r (ordinal i)
+    log1p (-exp lcdf)
+  quantile = \(Poisson r) q.
+    yield_state (0::Nat) \x.
+      with_state 0.0 \cdf.
+        while do
+          cdf := (get cdf) + exp (poisson_log_pdf_private r (get x))
+          if (get cdf) > q
+            then
+              False
+            else
+              x := (get x) + 1
+              True
+  draw = \(Poisson r) k.
+    exp_neg_rate = exp (-r)
+    [current_k, next_k] = split_key k
+    yield_state 0 \ans.
+      with_state (rand current_k) \p. with_state next_k \k'.
+        while do
+          if get p > exp_neg_rate
+            then
+              [ck, nk] = split_key (get k')
+              p := (get p) * rand ck
+              ans := (get ans) + 1
+              k' := nk
+              True
+            else
+              False
+
+-- TODO: remove "poisson" from prelude
+  
+
+'### Uniform distribution
+The [uniform distribution](https://en.wikipedia.org/wiki/Continuous_uniform_distribution) is continuous on an interval determined by a lower bound, `l`, and upper bound, `u > l`. Mean is `(l+u)/2`.
+
+data UniformDist = Uniform Float Float
+
+instance Dist UniformDist Float
+  log_pdf = \(Uniform l u) x.
+    if (x >= l)&&(x <= u)
+      then -log (u - l)
+      else -infinity
+  log_cdf = \(Uniform l u) x.
+    if (x < l)
+      then -infinity
+      else if (x > u)
+        then 0.0
+        else (log (x-l)) - (log (u-l))
+  log_sf = \(Uniform l u) x.
+    if (x < l)
+      then 0.0
+      else if (x > u)
+        then -infinity
+        else (log (u-x)) - (log (u-l))
+  quantile = \(Uniform l u) q.
+    l + ((u - l) * q)
+  draw = \(Uniform l u) k.
+    l + ((u - l) * (rand k))
+  
+
+'## Data summaries
+
+def mean_and_var {n} (xs:n=>Float) : (Float & Float) =
```

Comment:
I assumed that this function is here, because it will be faster to compute both mean and the var in a single pass, but it doesn't seem like it? Wouldn't it be simpler to just have a `mean` and a `var`?

Reply:
We could also generalize the types of right hand sides to be arbitrary vector spaces, which would later let us support vector-valued distributions etc.

Reply:
(what I mean is using `def var {n a} [VSpace a] (xs:n=>a) : a = ...`)

Reply:
Variance also requires `[Mul a]`.  That might be a reason to have a separate `mean` function.